fms/models/siglip_vision.py

```diff
@@ -0,0 +1,399 @@
+            emb_kq,
+            emb_v,
+            nheads,
+            kvheads,
+            p_dropout=self.config.attention_dropout,
+            use_bias=True,
+            linear_config=self.config.linear_config,
+            scale_factor=attn_scale_factor,
+        )
+
+        self.layer_norm2 = LayerNormParameterized(
+            self.embed_dim,
+            elementwise_shift=True,
+            use_mean=True,
+            eps=config.layer_norm_eps,
+            use_high_precision_pow=True,
+        )
+
+        self.mlp = FeedForwardBlock(
+            config.hidden_size,
+            hidden_grow_factor=config.intermediate_size // config.hidden_size,
+            activation_fn=str_to_activation(
+                config.hidden_act
+            ),  # NOTE: using nn.GELU as opposed to nn.functional.gelu() as in HF impl
+            use_bias=True,
+            linear_config=self.config.linear_config,
+            p_dropout=self.config.attention_dropout,
+        )
+
+    def reset_parameters(self):
+        for m in self.modules():
+            if (
+                isinstance(m, MultiHeadAttention)
+                or isinstance(m, LayerNormParameterized)
+                or isinstance(m, FeedForwardBlock)
+            ):
+                m.reset_parameters()
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        **attn_kwargs: Unpack[AttentionKwargs],
+    ):
+        attn_kwargs["attn_name"] = attn_kwargs.get("attn_name", "sdpa_bidirectional")
+
+        residual = hidden_states
+        hidden_states = self.layer_norm1(hidden_states)
+        hidden_states = self.self_attn(q=hidden_states, **attn_kwargs)
+        hidden_states = residual + hidden_states
+
+        residual = hidden_states
+        hidden_states = self.layer_norm2(hidden_states)
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = residual + hidden_states
+
+        return hidden_states
+
+
+class SiglipEncoder(nn.Module):
+    def __init__(self, config: SiglipVisionConfig):
+        super().__init__()
+        self.config = config
+        self.layers = nn.ModuleList(
+            [SiglipEncoderLayer(config) for _ in range(config.nlayers)]
+        )
+
+    def reset_parameters(self):
+        for m in self.layers:
+            m.reset_parameters()
+
+    def forward(
+        self,
+        inputs_embeds,
+        **attn_kwargs: Unpack[AttentionKwargs],
+    ):
+        hidden_states = inputs_embeds
+        for encoder_layer in self.layers:
+            hidden_states = encoder_layer(hidden_states, **attn_kwargs)
+        return hidden_states
+
+
+class SiglipMultiheadAttentionPoolingHead(nn.Module):
+    def __init__(self, config: SiglipVisionConfig):
+        super().__init__()
+        self.probe = nn.Parameter(torch.randn(1, 1, config.hidden_size))
+
+        # HF implementation uses PT MHA here, as opposed to SiglipAttnention as in the SiglipEncoderLayer
+        self.attention = torch.nn.MultiheadAttention(
```

Comment:
thinking mostly of AIU support for this, do we want to use our attention implementation instead?

Reply:
I tried using our MultiHeadAttention implementation, but lead to the error I shared before. Does using torch.nn variant lead to issues on AIU?

Reply:
ah, I see what's happening, this uses cross-attention, which our Attention class cannot do (we removed it as no models used cross attention anymore after everyone moved to decoders from encoder-decoder architectures). Keep it as torch.nn.MHA for now then, although I'm 99% sure this will fail on the AIU

Reply:
ouch 😅